### ISSUE TITLE
PDB not available

### ISSUE THREAD

**Issue by FilippoPolo** [microsoft/binskim#299]
Hi all,

I'm using a library that comes in a nuget package which doesn't include the PDB.

What would the best practice be in this case?

**Comment by scovetta** [comment]
As the package doesn't include symbols, they *might* be available via the website -- e.g.:
![image](https://user-images.githubusercontent.com/732166/98598469-5c59f880-228f-11eb-8749-a0090f50226b.png)

I'll leave it to others to comment on which checks require PDBs and which can be run without.

**Comment by FilippoPolo** [comment]
Hmm. Is there any way specify to binskim that it's acceptable to skip PDB-related checks for a given file?

**Comment by michaelcfanning** [comment]
Below is some information on this rule. There is no mechanism to skip PDB checks for a given file. For this case, the analysis itself remains incomplete and so the errors persist to the log file.

What we do internally for this is to produce a log file to serve as a baseline. Every time we analyze, we compare the new log to the baseline and only flag new errors. In this way, we don't lose visibility on the fact that we haven't completely vetted all dependencies. But neither do we see this condition raised over and over.

We probably should add a feature where users can pass in the *old* log file and where BinSkim will only mark new issues as new. Users might want a breaking mechanism as well, i.e., 'break on any newly introduced errors'.

Michael

https://github.com/microsoft/binskim/blob/master/docs/RulesAndErrorsTroubleshootingGuide.md

**Comment by quasarea** [comment]
Hi, @michaelcfanning  perhaps you could advise on error ERR997.ExceptionLoadingPdb related to that.

We are running binskim in azure devops pipelines and that breaks the pipeline and does not produce report:
```
Analysis did not complete due to one or more unrecoverable execution conditions.
Unexpected fatal runtime condition(s) observed: ExceptionLoadingPdb
```

I'm fine with knowing that I need to deliver pdb that are missing, but I would love to get report anyway :)

It was already mentioned in https://github.com/Microsoft/binskim/issues/161 but as the issue is quite old it seems orphaned.

**Comment by michaelcfanning** [comment]
Hello! #160 *is* an old issue, looks like Everett reactivated it. :)

This is a long-standing concern in running BinSkim for which there isn't an obvious solution that removes all user-friction. Product dependencies are part of a product's risk profile, so it is important to analyze these binaries in order to satisfy certain security policies (such as Microsoft's SDL). We haven't provided an 'ignore PDB load errors' knob in BinSkim to avoid the possibility that teams introduce this and compromise their security control. This is something that an auditor would object to if discovered ('you say that you analyze all shipping binaries for compiler-related security settings but I see that you've disabled some portion of analysis in your automated systems'). Instead, BinSkim persistently raises these gaps as errors in the output. Users can then resolve each error on its own terms, either by working to obtain the symbols or by ignoring the problem.

User can pass --rich-return-code on the command-line to receive a set or error/warning bits from the tool (rather than 0 - success, 1 - failure). One of these bits is ExceptionLoadingPdb (which has a value of 32/0x20). You can update your scripts to test for this bit on return and to ignore it (though I don't recommend it!).

A second option, if binaries don't need to be analyzed *at all* is to refine your automation scripts so that they aren't, in fact, analyzed.

The solution I've always had in mind for this issue is to improve BinSkim's base-lining. With this approach, users would run a single time and produce a baseline (which contains the PDB load errors). We would provide a way to subsequently run, passing in the log file and configuring BinSkim to only fail out in the event that *new* errors are introduced. With this approach, the current PDB errors are public (they are in the log file) but ongoing runs wouldn't persistently fail.

@eddynaka, I think it's time to actually implement this. :) Let's talk

@quasarea, would this be a workable solution from your perspective? If not, what are your other ideas?

**Comment by quasarea** [comment]
@michaelcfanning yes, makes sense, I could store this 'accepted-fails-report' along source code in repo, and when running binskim provide it to distinguish known issues from newly introduced and fail only on latter while report on both

**Comment by michaelcfanning** [comment]
Yes. I opened an [issue](https://github.com/microsoft/sarif-sdk/issues/2370) in the SARIF-SDK repo and we'll get started on it right away.

@eddynaka @yongyan-gh @shaopeng-gh

**Comment by eddynaka** [comment]
I just updated the sarif-sdk which contains the baseline option.
with this, we will be able to accomplish this.

**Comment by marekaldorf** [comment]
Closing issue thanks to work on sarif-sdk.
